snakemake/executors/slurm/slurm_jobstep.py

```diff
@@ -0,0 +1,117 @@
+import sys
+import time
+import shutil
+import tarfile
+import tempfile
+
+from snakemake.jobs import Job
+from snakemake.shell import shell
+from snakemake.logging import logger
+from snakemake.exceptions import print_exception
+from snakemake.exceptions import log_verbose_traceback
+from snakemake.exceptions import WorkflowError
+from snakemake.executors import ClusterExecutor
+from snakemake.utils import makedirs
+from snakemake.io import get_wildcard_names, Wildcards
+
+
+class SlurmJobstepExecutor(ClusterExecutor):
+    """
+    executes SLURM jobsteps and is *only* instaniated in
+    a SLURM job context
+    """
+
+    def __init__(
+        self,
+        workflow,
+        dag,
+        jobname="snakejob.{name}.{jobid}.sh",
+        printreason=False,
+        quiet=False,
+        printshellcmds=False,
+        latency_wait=3,
+        cluster_config=None,
+        local_input=None,
+        restart_times=None,
+        exec_job=None,
+        max_status_checks_per_second=10,
+    ):
+
+        # overwrite the command to execute a single snakemake job if necessary
+        # exec_job = "..."
+
+        super().__init__(
+            workflow,
+            dag,
+            None,
+            jobname=jobname,
+            printreason=printreason,
+            quiet=quiet,
+            printshellcmds=printshellcmds,
+            latency_wait=latency_wait,
+            cluster_config=cluster_config,
+            local_input=local_input,
+            restart_times=restart_times,
+            exec_job=exec_job,
+            assume_shared_fs=False,
+            max_status_checks_per_second=max_status_checks_per_second,
+        )
+        self.mem_per_node = os.getenv("SLURM_MEM_PER_NODE")
+        self.cpus_on_node = os.getenv("SLURM_CPUS_ON_NODE")
+        self.jobid = os.getenv("SLURM_JOB_ID")
+
+    def format_job(self, job):
+        return super().self.format_job(
+            self.exec_job, job, _quote_all=True, use_threads=True
+        )
+
+    def run(self, job, callback=None, submit_callback=None, error_callback=None):
+        super()._run(job)
+
+        jobsteps = dict()
+
+        if job.is_group():
+
+            def get_call(level_job, level_id, aux=""):
+                # we need this calculation, because of srun's greediness and
+                # SLURM's limits: it is not able to limit the memory if not per cpu
+                mem_per_cpu = level_job.resources.mem_mb // level_job.threads
+                return (
+                    "srun -J {level_id} --jobid {jobid}"
+                    "--mem-per-cpu {mem_per_cpu} -c {level_job.threads}"
+                    "--exclusive -n 1 {aux} {exec_job}".format(
+                        level_id=level_id,
+                        level_job=level_job,
+                        aux=aux,
+                        jobid=self.jobid,
+                        exec_job=self.format_job(level_job),
+                    )
+                )
+
+            for level in job.toposorted:
```

Comment:
This is quite cool, however it also means that we always wait until one level is completed, right? In principle, this can mean that resources are wasted.

Reply:
This entire thing runs on a single node, doesn't it?

Reply:
We definitely need to ensure that works with pipe producer and consumer jobs (job.toposorted should return them on the same level, but I think I have not yet ensured that).

Reply:
I have now ensured that in PR #1044.

Reply:
I wonder if it's possible to use `srun`'s `--dependency` to manage submission dependency, rather than toposort layers. My idea is you would iterate over jobs in toposort order, but keep track of their step id (`<jobid>.<step>` where step increments from 0...). If a job has dependencies according to the DAG, you would communicate that to slurm using the step ids we've been tracking. Then slurm would start each job as soon as each dependency finishes, preventing "one layer at a time" resource wastage.